homeassistant/helpers/script.py

```diff
@@ -485,26 +513,203 @@ def async_script_timeout(now):
+            action = _determine_action(step)
+
+            if action == ACTION_CHECK_CONDITION:
+                referenced |= condition.async_extract_devices(step)
+
+            elif action == ACTION_DEVICE_AUTOMATION:
+                referenced.add(step[CONF_DEVICE_ID])
+
+        self._referenced_devices = referenced
+        return referenced
+
+    @property
+    def referenced_entities(self):
+        """Return a set of referenced entities."""
+        if self._referenced_entities is not None:
+            return self._referenced_entities
+
+        referenced = set()
+
+        for step in self.sequence:
+            action = _determine_action(step)
+
+            if action == ACTION_CALL_SERVICE:
+                data = step.get(service.CONF_SERVICE_DATA)
+                if not data:
+                    continue
+
+                entity_ids = data.get(ATTR_ENTITY_ID)
+
+                if entity_ids is None:
+                    continue
+
+                if isinstance(entity_ids, str):
+                    entity_ids = [entity_ids]
+
+                for entity_id in entity_ids:
+                    referenced.add(entity_id)
+
+            elif action == ACTION_CHECK_CONDITION:
+                referenced |= condition.async_extract_entities(step)
+
+            elif action == ACTION_ACTIVATE_SCENE:
+                referenced.add(step[CONF_SCENE])
+
+        self._referenced_entities = referenced
+        return referenced
+
+    def run(self, variables=None, context=None):
+        """Run script."""
+        asyncio.run_coroutine_threadsafe(
+            self.async_run(variables, context), self._hass.loop
+        ).result()
+
+    async def async_run(
+        self, variables: Optional[Sequence] = None, context: Optional[Context] = None
+    ) -> bool:
+        """Run script."""
+        if self.is_running:
+            if self._if_running == IF_RUNNING_IGNORE:
+                self._log("Skipping script")
+                return False
+            if self._if_running == IF_RUNNING_ERROR:
+                self._raise("Already running")
+            if self._if_running == IF_RUNNING_RESTART:
+                self._log("Restarting script")
+                self.async_stop()
+
+        self.last_triggered = utcnow()
+        if self._run_mode == RUN_MODE_LEGACY:
+            if self.is_running:
+                shared = cast(Optional[_LegacyScriptRun], self._runs[0])
+            else:
+                shared = None
+            run = cast(
+                _ScriptRunBase,
+                _LegacyScriptRun(self._hass, self, variables, context, shared),
+            )
+            self._runs.append(run)
+            await run.async_run()
+        else:
+            run = cast(_ScriptRunBase, _ScriptRun(self._hass, self, variables, context))
+            self._runs.append(run)
+            cast(_ScriptRun, run).task = self._hass.async_create_task(run.async_run())
+            if self._run_mode == RUN_MODE_BLOCKING:
```

Comment:
I don't understand this, I would expect this logic to be in the `_ScriptRun` inherited classes. For `_BlockingScriptRun` it would just run it, while in `_AsynchronousScriptRun` the `async_run` method would schedule it and store the task.

Reply:
Well, the way I looked at it, _what_ `_ScriptRun.async_run` did was independent of whether or not the caller waited for it to finish. That's why I implemented it the way I did. But due to not being able to find `asyncio.current_task`, the implementation was probably more complicated than it needed to be (and used one extra Task in the blocking case than it needed to.) Again, let me work on it...

Reply:
When I first read your code, I was totally expecting different `_ScriptRun` inherited classes so you can handle the running differently in `async_run`. I understand now that the base class was only to have a new one just for legacy behavior. I think that we could still be cleaner if we have different classes. Also would make testing easier as it's easy to test those different classes independently.

Reply:
Yeah, good point. I hadn't thought of it. I'll add it to the list for the re-work & next commit. Thanks!